packages/make-hardener/src/harden-branding.js

```diff
@@ -0,0 +1,190 @@
+    // @ts-ignore
+    ThrowTypeError: getOwnPropertyDescriptor(Function.prototype, 'caller').get,
+    SafeMap,
+    SafeMapProto,
+  };
+
+  for (const v of values(possiblyAlreadyNonExtensible)) {
+    if (!isPrimitive(v) && !isExtensible(v)) {
+      // @ts-ignore
+      onlyNonExtensible.add(v);
+    }
+  }
+
+  const patchObjectMethods = {
+    freeze(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
+      }
+      return originalFreeze(target);
+    },
+    seal(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
+      }
+      return originalSeal(target);
+    },
+    preventExtensions(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
+      }
+      return originalObjectPreventExtensions(target);
+    },
+  };
+
+  const patchReflectMethods = {
+    // Note that Object.preventExtensions has a slightly different
+    // semantics than Reflect.preventExtensions
+    preventExtensions(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
+      }
+      return originalReflectPreventExtensions(target);
+    },
+  };
+
+  defineProperties(
+    Object,
+    Object.getOwnPropertyDescriptors(patchObjectMethods),
+  );
+
+  defineProperties(
+    Reflect,
+    Object.getOwnPropertyDescriptors(patchReflectMethods),
+  );
+}
+
+/**
+ * Currently not exported outside the package.
+ * TODO But we should consider exporting it, after removing the `path`
+ * parameter.
+ *
+ * @param {any} val
+ * @param {string=} path
+ * @returns {boolean}
+ */
+export const isHardened = (val, path) => {
+  if (isPrimitive(val)) {
+    return true;
+  }
+  let answer1;
+  let answer2;
+  if (ENABLE_POSITIVE_BRANDING) {
+    answer1 = hardened.has(val);
+  }
+  if (ENABLE_NEGATIVE_BRANDING) {
+    // We check isFrozen, but we accumulate objects when they become
+    // non-extensible, because once non-extensible they might become frozen,
+    // and we can more feasibly catch all operations that might make them
+    // non-extensible.
+    answer2 = isFrozen(val) && !onlyNonExtensible.has(val);
+  }
+  if (ENABLE_POSITIVE_BRANDING && ENABLE_NEGATIVE_BRANDING) {
+    if (answer1 !== answer2) {
+      // Not necessarily an error. Might be one of the three safe
+      // exceptions explained in the README.md. We should test these
+      // outliers, because if it is anything else, we may have a
+      // serious bug.
+      console.log(`HARDEN disagreement: ${answer1} vs ${answer2} at ${path}`);
+    }
+  }
+
+  return !!(answer1 && answer2);
+};
+
+// TODO curried forEach
+// We capture the real WeakSet.prototype.add above, in case someone
+// changes it. The two-argument form of forEach passes the second
+// argument as the 'this' binding, so we add to the correct set.
+export const moreObjectsHardened = toHarden => {
+  if (ENABLE_POSITIVE_BRANDING && ENABLE_NEGATIVE_BRANDING) {
+    // TODO What consistency check should we put here?
```

Comment:
TODO👆 I presume the intersection of hardened and onlyNotExtensible should be empty.